sound/soc/sof/ipc4-pcm.c

```diff
@@ -314,9 +315,99 @@ static int sof_ipc4_pcm_hw_params(struct snd_soc_component *component,
 	return sof_ipc4_build_time_info(sdev, &spcm->stream[substream->stream]);
 }
 
+static int sof_ipc4_get_stream_start_offset(struct snd_sof_dev *sdev,
+					    struct snd_pcm_substream *substream,
+					    struct snd_sof_pcm_stream *stream,
+					    struct sof_ipc4_timestamp_info *time_info)
+{
+	struct sof_ipc4_copier *host_copier = time_info->host_copier;
+	struct sof_ipc4_copier *dai_copier = time_info->dai_copier;
+	struct sof_ipc4_pipeline_registers ppl_reg;
+	u64 stream_start_position;
+	u32 dai_sample_size;
+	u32 ch, node_id;
+	u32 offset;
+
+	node_id = SOF_IPC4_NODE_INDEX(host_copier->data.gtw_cfg.node_id);
+	offset = offsetof(struct sof_ipc4_fw_registers, pipeline_regs) + node_id * sizeof(ppl_reg);
+	sof_mailbox_read(sdev, sdev->fw_info_box.offset + offset, &ppl_reg, sizeof(ppl_reg));
+	if (ppl_reg.stream_start_offset == SOF_IPC4_INVALID_STREAM_POSITION)
+		return -EBUSY;
+
+	stream_start_position = ppl_reg.stream_start_offset;
+	ch = dai_copier->data.out_format.fmt_cfg;
+	ch = SOF_IPC4_AUDIO_FORMAT_CFG_CHANNELS_COUNT(ch);
+	dai_sample_size = (dai_copier->data.out_format.bit_depth >> 3) * ch;
+	/* convert offset to sample count */
+	do_div(stream_start_position, dai_sample_size);
+	time_info->stream_start_offset = stream_start_position;
+
+	return 0;
+}
+
+static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
+					    struct snd_pcm_substream *substream)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct sof_ipc4_timestamp_info *time_info;
+	struct sof_ipc4_llp_reading_slot llp;
+	struct snd_sof_pcm_stream *stream;
+	struct snd_sof_pcm *spcm;
+	snd_pcm_uframes_t host_ptr;
+	u64 link_ptr, tmp_ptr;
+	int ret;
+
+	spcm = snd_sof_find_spcm_dai(component, rtd);
+	if (!spcm)
+		return 0;
+
+	stream = &spcm->stream[substream->stream];
+	time_info = stream->private;
+	if (!time_info)
+		return 0;
+
+	/*
+	 * stream_start_offset is updated to memory window by FW based on
+	 * pipeline statistics and it may be invalid if host query happens before
+	 * the statistics is complete. And it will not change after the first initiailization.
+	 */
+	if (time_info->stream_start_offset == SOF_IPC4_INVALID_STREAM_POSITION) {
+		ret = sof_ipc4_get_stream_start_offset(sdev, substream, stream, time_info);
+		if (ret < 0)
+			return 0;
+	}
+
+	/* Intel HDA platform doesn't support llp counter based on GP-DMA */
+	if (!time_info->llp_offset) {
+		tmp_ptr = snd_sof_pcm_get_stream_position(sdev, component, substream);
+		if (!tmp_ptr)
+			return 0;
+	} else {
+		sof_mailbox_read(sdev, time_info->llp_offset, &llp, sizeof(llp));
+		tmp_ptr = ((u64)llp.reading.llp_l | ((u64)llp.reading.llp_u << 32));
+	}
+
+	/*
+	 * in rare case audio stream can last more than 24.855 hours
+	 * for 48khz 2ch case, do wrap to a 32 bit counter
+	 */
+	if (tmp_ptr > UINT_MAX) {
+		tmp_ptr -= 1;
+		link_ptr = do_div(tmp_ptr, UINT_MAX);
```

Comment:
Sorry, I may need more coffee but I couldn't figure out what you are trying to do here and what formula this implements?

Reply:
Definition:  u32 ret = do_div(u64, u32).   

If tmp_ptr > UINT_MAX,    .e.g. tmp_ptr  = UINT_MAX + 1,   
link_ptr = do_div(tmp_ptr, UINT_MAX) = 1, but it should be 0 since it is wrapped to 0 for ringbuffer

Reply:
if do_div can accept u64 for second params, it is easy to  do with do_div(tmp_ptr,  UINT_MAX+1)